_drafts/implementing-state-machines-in-postgresql.md

```diff
@@ -0,0 +1,301 @@
+process of coming up with: a set of states an order can be in, a set of events,
+a transition function that maps the combination of each state and event to a
+new state, and an initial state.
+
+In practice this is usually done by drawing up a directed graph that shows how
+the different states are connected to each other via events:
+
+<img src="{% asset_path orders.svg %}" alt="Finite-state machine for an order management system in PostgreSQL" />
+
+Besides guiding our implementation, these graphs can be very useful for
+discussions and analysis. For example you can immediately see how there is no
+way for a customer to return an order after it has shipped yet. Additionally
+the process of naming the involved states and events automatically creates a
+precise language that can be adopted by all stakeholders.
+
+If you paid close attention so far, you might have noticed that the graph above
+does not show the combination of all states and events which simply implies
+that the missing combinations lead to an error state. The [complete FSM]({%
+asset_path orders-full.svg %}) can be shown as a graph as well, but I don't
+think it's very useful in practice.
+
+Anyway, as promised in the title of this article we're going to implement this
+FSM in PostgreSQL. This may not be everybody's cup of tea, but you might like
+how this approach allows us gives advanced analytical powers as a free
+by-product. Embedding this kind of logic into the DB can also help protecting
+against race conditions, but this will perhaps be the topic of a future post.
+
+Let's begin by creating an `order_events` table which keeps track of all events
+for a given `order_id`.
+
+```sql
+CREATE TABLE order_events (
+  id serial PRIMARY KEY,
+  order_id int NOT NULL,
+  event text NOT NULL,
+  time timestamp DEFAULT now() NOT NULL
+);
+```
+
+Next, let's implement the transition function, which is the heart of every
+FSM:
+
+```sql
+CREATE FUNCTION order_events_transition(state text, event text) RETURNS text
+LANGUAGE sql AS
+$$
+  SELECT CASE state
+    WHEN 'start' THEN
+      CASE event
+        WHEN 'create' THEN 'awaiting_payment'
+        ELSE 'error'
+      END
+    WHEN 'awaiting_payment' THEN
+      CASE event
+        WHEN 'pay' THEN 'awaiting_shipment'
+        WHEN 'cancel' THEN 'canceled'
+        ELSE 'error'
+      END
+    WHEN 'awaiting_shipment' THEN
+      CASE event
+        WHEN 'cancel' THEN 'awaiting_refund'
+        WHEN 'ship' THEN 'shipped'
+        ELSE 'error'
+      END
+    WHEN 'awaiting_refund' THEN
+      CASE event
+        WHEN 'refund' THEN 'canceled'
+        ELSE 'error'
+      END
+    ELSE 'error'
+  END
+$$;
+```
+
+And before we proceed, let's check a few examples to make sure the function is
+working:
+
+```sql
+SELECT state, event, order_events_transition(state, event)
+FROM (VALUES
+  ('start', 'create'),
+  ('awaiting_payment', 'pay'),
+  ('awaiting_payment', 'cancel'),
+  ('awaiting_payment', 'ship')
+) examples(state, event);
```

Comment:
Ahhh, figured it out! :) Minor suggestions: I'd use `AS examples(state, event)` here to maybe not trip other SQL-Peasants up